core/dnsserver/server_quic.go

```diff
@@ -202,16 +219,20 @@ func (s *ServerQUIC) Serve(l net.Listener) error { return nil }
 func (s *ServerQUIC) Listen() (net.Listener, error) { return nil, nil }
 
 // closeQUICConn quietly closes the QUIC connection.
-func closeQUICConn(conn quic.Connection, code quic.ApplicationErrorCode) {
+func (s *ServerQUIC) closeQUICConn(conn quic.Connection, code quic.ApplicationErrorCode) {
 	if conn == nil {
 		return
 	}
 
 	clog.Debugf("closing quic conn %s with code %d", conn.LocalAddr(), code)
-
 	err := conn.CloseWithError(code, "")
 	if err != nil {
-		clog.Debugf("closing quic connection with code %d: %s", code, err)
+		clog.Debugf("failed to close quic connection with code %d: %s", code, err)
+	}
+
+	// DoQCodeNoError metrics are already registered after s.ServeDNS()
```

Comment:
This part is a little tricky: metrics of successful queries will be registered after ServeDNS. 

However, for failed responses, I chose to report their metrics after closing their connections. This is the way I found to have one response per stream.